### PR TITLE
route_manager

### PR DESCRIPTION
**Pull request by Inno97**
Create the RouteManager which assists the user in manipulating routes. The RouteManager will 
"select" a Route or RouteNode, which lets the user interact without having to type much.

Features:

	 - Rename RouteShow to RouteList
		 - RouteList will list the information
		 - RouteShow will display the route in map

	 - Add RouteShow command
		 - Show Routes and their RouteNodes in Singapore map
			 - Show points on map
				 - Differentiate starting and ending RouteNodes
			 - Show name on map
			 - Show description on hover
			 - Highlight route on map
			 - Unforuntately, it'll take too long for me to make the map smaller and more readable,
			 - So this features is a little redundant.
	 
	 - Added RouteNodeShow command
		 - Shows the RouteNode and all nearby RouteNodes via staticmap
		
	
	 - Add mapping feature to route
		 - Display via command
	 
	 - Added RouteManager
		 - Middleman class that assists in route manipulation
			 - Stores the index of the current route and/or node that is being edited
			 - Automatically fills in information for commands according to the above
			 - Functions mostly as conversations
		 
		 - Implementation
			 - Is turned on and off via commands
			 - If no conversation is ongoing, RouteManager will append a prefix to any input
				 - This prefix with the input, will trigger the proper conversation or command
				 - Conversation moves on as normal, but with reduced number of fields to fill in
				 - Upon ending conversation, RouteManager compiles result and returns it
			 - Most conversations will be internally split into 2 branches
				 - 2 to support both Route and RouteNode commands
				 
		 - Functions / Conversation starters
			 - select
				 - Makes a conversation to select the route via index
			 - node
				 - Makes a conversation to select the node via index
				 - Will reject based on whether the current route has such a node
				 
			 - returns
				 - Selects back the route
			 
			 - edit
				 - Makes a conversation to select the field and new value

			 - add
				 - Route
					 - Makes a conversation to add a new Route normally
					 - Selects this Route upon creation
			 
				 - RouteNode
					 - Fetches the local data of the node, to show nearby nodes
					 - Also fetches the StaticMap image
					 - Makes a conversation as normal to add a new RouteNode
					 - Selects this RouteNode upon creation
					 - Adds the new node infront of the selected node	 
			 - info
				 - Route
					 - Displays all the Routes and RouteNodes as normal
				 
				 - RouteNode
					 - Displays the information of the RouteNode
						 - Show any transport-related information if possible
						 - Fetches the StaticMap image of the RouteNode
						 
			 - generate
				 - Makes a conversation to generate a Route normally
					 - Selects this Route upon creation
					 - Fetches the StaticMap image of the starting node
			 - help
				 - Returns the help message for the RouteManager in the given context
				 
		 - Add fetching of nearby nodes support
			 - For purposes of StaticMap visiblity, max limit is 3km
			 
			 - nearby
				 - RouteNode
					 - Displays some nearby BusStops and TrainStations
					 - Fetches StaticMap and highlight nearby nodes

### CODE REVIEW

**Comment by Inno97** [comment]
RouteManager still has quite a lot of bugs, I will be working on fixing it in v1.4.
All essential functions are implemented, but it requires quite a lot of cleanup.

File: src/main/java/duke/logic/LogicManager.java [hunk 1]
```diff
@@ -44,7 +44,10 @@ public CommandResult execute(String userInput) throws DukeException {
         if (EditorManager.isActive()) {
             logger.log(Level.INFO, "editing...");
             c = EditorManager.edit(userInput);
-        } else {
+        } else  {
+            if (model.getRouteManager().isActivated() && !conversationManager.isInConversation()) {
```
Comment (Jefferson111) [inline src/main/java/duke/logic/LogicManager.java]: LOD

File: src/main/java/duke/logic/api/ApiParser.java [hunk 2]
```diff
@@ -173,21 +173,22 @@ public static String generateStaticMapPolygon(ArrayList<String> points, String r
...
+            result = result.substring(0, result.length() - 1) + "]";
 
-        if (!rgb.isEmpty()) {
-            result += ":" + rgb;
-        } else {
-            result += ":0,0,0";
+            if (!rgb.isEmpty()) {
+                result += ":" + rgb + ":" + lineWidth;
+            } else {
+                result += ":0,0,0:" + lineWidth;
```
Comment (Jefferson111) [inline src/main/java/duke/logic/api/ApiParser.java]: use constant

File: src/main/java/duke/logic/commands/RouteManagerHelpCommand.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+    /**
+     * Executes this command and returns a text result.
+     *
+     * @param model The model object containing event list.
+     */
+    @Override
+    public CommandResultText execute(Model model) {
+
+        return new CommandResultText(model.getRouteManager().getHelpMessage());
```
Comment (Jefferson111) [inline src/main/java/duke/logic/commands/RouteManagerHelpCommand.java]: LOD

File: src/main/java/duke/logic/commands/RouteNodeNeighboursCommand.java [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+        }
+
+        for (Map.Entry mapElement : model.getMap().getTrainMap().entrySet()) {
+            if (((TrainStation) mapElement.getValue()).getDistance(node) < DISTANCE_THRESHOLD
+                    && ((TrainStation) mapElement.getValue()).getAddress() != node.getAddress()) {
+                temp.add((TrainStation) mapElement.getValue());
+            }
+        }
+
+        temp.sort((Venue o1, Venue o2) -> (int) (o1.getDistance(node) - o2.getDistance(node)));
```
Comment (Jefferson111) [inline src/main/java/duke/logic/commands/RouteNodeNeighboursCommand.java]: Use VenueList and add sort function to the list.

File: src/main/java/duke/logic/commands/RouteNodeNeighboursCommand.java [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+    /**
+     * Generates 6 other nodes beside the selected one in the Route as an ArrayList.
+     *
+     * @param route The Route object.
+     * @param query The original RouteNode being queried.
+     * @return points The ArrayList of points.
+     */
+    public ArrayList<String> generateOtherPoints(Route route, RouteNode query) {
+        ArrayList<String> points = new ArrayList<>();
```
Comment (Jefferson111) [inline src/main/java/duke/logic/commands/RouteNodeNeighboursCommand.java]: String builder

**Pull request merged**